cmd/crank/beta/validate/validate.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/crossplane/crossplane/internal/controller/apiextensions/composite"
+	"github.com/crossplane/crossplane/internal/xcrd"
+)
+
+func convertToCRDs(schemas []*unstructured.Unstructured) ([]*extv1.CustomResourceDefinition, error) { //nolint:gocyclo // Not a complex function, just switch/case statements
+	crds := make([]*extv1.CustomResourceDefinition, 0, len(schemas))
+	for _, s := range schemas {
+		switch s.GroupVersionKind().GroupKind() {
+		case schema.GroupKind{Group: "apiextensions.k8s.io", Kind: "CustomResourceDefinition"}:
+			crd := &extv1.CustomResourceDefinition{}
+			bytes, err := s.MarshalJSON()
+			if err != nil {
+				return nil, errors.Wrap(err, "cannot marshal CRD to JSON")
+			}
+
+			if err := yaml.Unmarshal(bytes, crd); err != nil {
+				return nil, errors.Wrap(err, "cannot unmarshal CRD YAML")
+			}
+
+			crds = append(crds, crd)
+
+		case schema.GroupKind{Group: "apiextensions.crossplane.io", Kind: "CompositeResourceDefinition"}:
+			xrd := &v1.CompositeResourceDefinition{}
+			bytes, err := s.MarshalJSON()
+			if err != nil {
+				return nil, errors.Wrap(err, "cannot marshal XRD to JSON")
+			}
+
+			if err := yaml.Unmarshal(bytes, xrd); err != nil {
+				return nil, errors.Wrap(err, "cannot unmarshal XRD YAML")
+			}
+
+			crd, err := xcrd.ForCompositeResource(xrd)
+			if err != nil {
+				return nil, errors.Wrapf(err, "cannot derive composite CRD from XRD %q", xrd.GetName())
+			}
+			crds = append(crds, crd)
+
+			if xrd.Spec.ClaimNames != nil {
+				claimCrd, err := xcrd.ForCompositeResourceClaim(xrd)
+				if err != nil {
+					return nil, errors.Wrapf(err, "cannot derive claim CRD from XRD %q", xrd.GetName())
+				}
+				crds = append(crds, claimCrd)
+			}
+
+		case schema.GroupKind{Group: "pkg.crossplane.io", Kind: "Provider"}:
+			fmt.Println("Provider extension is not supported yet")
+			continue
+
+		case schema.GroupKind{Group: "pkg.crossplane.io", Kind: "Configuration"}:
+			fmt.Println("Configuration extension is not supported yet")
+			continue
+
+		default:
+			continue
+		}
+	}
+
+	return crds, nil
+}
+
+func newValidators(crds []*extv1.CustomResourceDefinition) (map[schema.GroupVersionKind]validation.SchemaValidator, error) {
+	validators := map[schema.GroupVersionKind]validation.SchemaValidator{}
+
+	for i := range crds {
+		internal := &ext.CustomResourceDefinition{}
+		if err := extv1.Convert_v1_CustomResourceDefinition_To_apiextensions_CustomResourceDefinition(crds[i], internal, nil); err != nil {
+			return nil, err
+		}
+
+		for _, ver := range internal.Spec.Versions {
+			var sv validation.SchemaValidator
+			var err error
+
+			if ver.Schema != nil && ver.Schema.OpenAPIV3Schema != nil {
+				sv, _, err = validation.NewSchemaValidator(ver.Schema.OpenAPIV3Schema)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			if internal.Spec.Validation != nil {
+				sv, _, err = validation.NewSchemaValidator(internal.Spec.Validation.OpenAPIV3Schema)
```

Comment:
is this implying that we'll prefer the validation from `internal.Spec.Validation` over that of `ver.Schema` and `ver.Schema.OpenAPIV3Schema`? there's no need to use all of them?

This may be another instance where a few comments explaining the choices we're making in easy plain english as we're going along would be helpful for future readers 🤓

Reply:
Thanks for the question @jbw976! After this comment I double checked the logic and looks like top level validation rules and version based validations are [mutually exclusive](https://github.com/kubernetes/apiextensions-apiserver/blob/80788265e19419eafebe95c670234eb2659bc072/pkg/apis/apiextensions/types.go#L50). I updated the code to use both of them.

Reply:
this is on the API Server internal representation of CRDs, CRDs don't have a top level schema at all, as you can see [here](https://github.com/kubernetes/apiextensions-apiserver/blob/80788265e19419eafebe95c670234eb2659bc072/pkg/apis/apiextensions/v1/conversion.go#L149), the conversion code sets the top level schema if all the per-version ones are identical, so it's safe to use the top-level one if available, so you could first check that and continue if found actually.